src/main/java/algorithms/minimumSpanningTree/prims/Prim.java

```diff
@@ -1,115 +1,80 @@
-        PriorityQueue<Pair> pq = new PriorityQueue<>(Comparator.comparingInt(Pair::getDist));
-        for (int i = 0; i < v; i++) {
-            if (!mst.contains(i)) {
-                if (adjM[0][i] != 0) { // ensure valid edge
-                    pq.add(new Pair(adjM[0][i], i));
-                }
-            }
+    public static List<Node> getPrimsMST(List<Node> graph) {
+        PriorityQueue<Node> pq = new PriorityQueue<>((a, b) -> a.getCurrMinWeight() - b.getCurrMinWeight());
+
+        // Values in the map represent the corresponding node with only the edges in the MST
+        Map<Node, Node> nodeToMSTNode = new HashMap<>();
+        Map<Node, Node> parentInMST = new HashMap<>();
+
+        // Initialize each node's minWeight to infinity and add to the priority queue
+        for (Node node : graph) {
+            node.setCurrMinWeight(Integer.MAX_VALUE);
+            pq.add(node);
+            nodeToMSTNode.put(node, new Node(node.getIdentifier())); // Create a corresponding MST node
+            parentInMST.put(node, null);
         }
-        while (mst.size() != v) {
-            Pair popped = pq.poll();
-            if (mst.contains(popped.getEndNode())) {
-                continue;
-            }
-            minCost += popped.getDist();
-            mst.add(popped.getEndNode());
-            for (int i = 0; i < v; i++) {
-                if (mst.contains(i)) {
-                    continue;
-                }
-                if (adjM[popped.getEndNode()][i] != 0) { // ensure valid edge
-                    pq.add(new Pair(adjM[popped.getEndNode()][i], i));
-                }
-            }
+
+        // Assuming graph is not empty and the start node is the first node
+        if (!graph.isEmpty()) {
+            graph.get(0).setCurrMinWeight(0);
         }
-        return minCost;
-    }
 
-    /**
-     * Alternative implementation that simply uses array to hold weights rather than heap.
-     * Note: Starts from the node labelled 0 and repeatedly update weights
-     * which stores the minimum weight from any node in the MST to other nodes.
-     * Time: O(V) + O(V*2V)
-     * Space: O(V)
-     *
-     * @param adjM Adjacency matrix that encapsulates the distance/weight between nodes
-     * @return minimum weight of the spanning tree
-     */
-    public int minCostConnectPoints2(int[][] adjM) {
-        int v = adjM.length;
-        int[] weights = new int[v];
+        while (!pq.isEmpty()) {
+            Node current = pq.poll();
+            current.setVisited(true);
+
+            Map<Node, Integer> currentAdjacentNodes = current.getAdjacentNodes();
+
+            for (Map.Entry<Node, Integer> entry : currentAdjacentNodes.entrySet()) {
+                Node adjacent = entry.getKey();
+                Integer weight = entry.getValue();
 
-        System.arraycopy(adjM[0], 0, weights, 0, v);
+                if (!adjacent.isVisited() && weight < adjacent.getCurrMinWeight()) {
+                    pq.remove(adjacent);
```

Comment:
remove() is actually not a typical operation supported by heap / PriorityQueue (see the GitHub notes on Heap).

While java implementation includes a remove operation (for convenience), this remove will ultimately incur O(n) since a heap (if not augmented), can do no better but O(n) to first search for the node before proceeding with removal.